### PR TITLE
br: refactor br cli backoff logic

### LINKED ISSUE #54643: br: refactor CLI backoff retry logic
## Enhancement
The current BR CLI backoff retry has some redundant logic especially in the `NextBackoff`. every time creating a new rule needs to impl the `NextBackoff`. We can probably have one single `NextBackoff` impl and have errors to retry passed in as set etc. Make it more simplified and easy to extend in the future.

### PR DESCRIPTION
**Pull request by Tristan1900**
### What problem does this PR solve?
Issue Number: close #54643 

Problem Summary:

Refactor the BR CLI backoff retry logic to make is more simplified and extendable in future.

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
NA
```

### CODE REVIEW

**Comment by Tristan1900** [comment]
tests to be fixed 
also pending on https://github.com/pingcap/tidb/pull/54596

File: br/pkg/utils/backoff.go [hunk 1]
```diff
@@ -128,204 +127,246 @@ func (rs *RetryState) ReduceRetry() {
...
+	nonRetryErrs := map[error]struct{}{
+		context.Canceled:                 {},
+		berrors.ErrKVRangeIsEmpty:        {},
+		berrors.ErrKVRewriteRuleNotFound: {},
+	}
+
+	isRetryErrFunc := buildIsRetryErrFunc(retryErrs, grpcRetryCodes)
+	isNonRetryErrFunc := buildIsNonRetryErrFunc(nonRetryErrs)
+
+	return newBackoffStrategy(maxRetry, delayTime, maxDelayTime, errContext, isRetryErrFunc, isNonRetryErrFunc)
```
Comment (YuJuncen) [inline br/pkg/utils/backoff.go:212]: Could you replace this constructor with functional options or builder pattern?

**Review by YuJuncen — APPROVED** [review]
rest lgtm

File: br/pkg/restore/log_client/import_retry.go [hunk 2]
```diff
@@ -251,10 +240,6 @@ func (r *RPCResult) StrategyForRetryStoreError() RetryStrategy {
 }
 
 func (r *RPCResult) StrategyForRetryGoError() RetryStrategy {
-	if r.Err == nil {
```
Comment (Tristan1900) [inline br/pkg/restore/log_client/import_retry.go:254]: only calls when error is not nil

**Review by Leavrth — APPROVED** [review]

**Review by GMHDBJD — APPROVED** [review]
LGTM for dumpling

**Pull request merged**